sequence_task_1/student/include/impl.h

```diff
@@ -1,51 +1,131 @@
-#pragma once
+#ifndef _SEQUANCE_TASK_1_
+#define _SEQUANCE_TASK_1_
+
 #include <string>
+#include <memory>
+#include <utility>
 
 struct Song
 {
     std::string name;
     explicit Song(std::string song_name)
-        : name {std::move(song_name)}
+        : name{std::move(song_name)}
     {
     }
 };
 
-bool operator==(const Song& lhs, const Song& rhs)
+bool operator==(const Song &lhs, const Song &rhs)
 {
     return lhs.name == rhs.name;
 }
 
-template<class Container>
+template <class Container>
 class StaticPlaylist
 {
 public:
     /** @todo Member traits */
 
+    typedef typename Container::size_type           size_type;
+    typedef typename Container::value_type          value_type;
+    typedef typename Container::reference           reference;
+    typedef typename Container::const_reference     const_reference ;
+    //typedef typename Container::type type;
+
     /** @todo Iterators */
+    typedef typename Container::iterator            iterator;
+    typedef typename Container::const_iterator      const_iterator;
+    typedef typename iterator::difference_type      difference_type;
+    typedef typename iterator::pointer              pointer;
+    typedef const pointer                           const_pointer;
+
+    const_iterator begin()
+    {
+        return m_tracklist.begin();
+    }
+
+    const_iterator end()
+    {
+        return std::end(m_tracklist);
+    }
+
+    const_iterator cbegin()
+    {
+        return m_tracklist.cbegin();
+    }
+
+    const_iterator cend()
+    {
+        return m_tracklist.cend();
+    }
 
     /** @todo Constructor from any reversible sequence container */
+   
+    template <class TPlaylist>
+    StaticPlaylist(const TPlaylist& other)
+        :m_tracklist()
+    {
+        m_tracklist.assign(other.begin(), other.end());
+    }
 
     /** @todo Assignment from any reversible sequence container */
+    template <class TPlaylist>
+    StaticPlaylist& operator=(TPlaylist other)
+    {
+        m_tracklist.clear();
+        m_tracklist.assign(other.begin(), other.end());
+        m_tracklist.shrink_to_fit();
+        return *this;
+    }
 
     /** @todo Add track from initializer */
-    template<class... Args>
-    const Song& play(Args&&... songData);
+    template <class... Args>
+    const Song &play(Args &&... songData)
+    {
+        size_type current_size = count();
+        m_tracklist.insert(m_tracklist.end(), {Song(songData)...});
```

Comment:
You don't need to construct a temporary object, if you already have the initialization data, that can be used for in-place construction.